### PR TITLE
arch/risc-v: Refine Toolchain.defs

### PR DESCRIPTION
**Pull request by no1wudi**
## Summary
Refine risc-v Toolchain.defs firstly, and it tested on qemu-rv32.
And then:
1. Apply new Toolchain.defs to all risc-v based chips 
2. Merge the common src for rv32/rv64 (riscv_syscall.S/riscv_fpu.S etc, in rv32im and rv64gc)
3. Apply new common src to qemu-rv32 and qemu-rv64
4. Migrate all risc-v based chips to new implementation and remove old
## Impact
Refactor only.
## Testing
Tested on qemu-rv32 and CI

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
Look like we should finish this PR: https://github.com/apache/incubator-nuttx/pull/2822

**Comment by no1wudi** [comment]
> Look like we should finish this PR: #2822

This PR will be finished in next phase, I will pick it later.

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/risc-v/Kconfig [hunk 1]
```diff
@@ -82,13 +106,19 @@ config ARCH_CHIP_MPFS
...
 	---help---
 		NXP RV32M1 processor (RISC-V Core with PULP extensions).
 
 config ARCH_CHIP_QEMU_RV32
 	bool "QEMU RV32"
-	select ARCH_RV32IM
+	select ARCH_RV32
+	select ARCH_RV_ISA_M
+	select ARCH_RV_ISA_C
+	select ARCH_RV32IM # TODO: Remove this old style options
```
Comment (xiaoxiang781216) [inline arch/risc-v/Kconfig:121]: why not remove now

File: arch/risc-v/src/common/Toolchain.defs [hunk 2]
```diff
@@ -53,15 +53,66 @@ else
...
+    ARCHABITYPE = lp64
+  endif
+
+  # Construct arch flags
+
+  ARCHCPUFLAGS = -march=$(ARCHTYPE)i$(ARCHRVISAM)$(ARCHRVISAA)$(ARCHRVISAF)$(ARCHRVISAD)$(ARCHRVISAC)
+
+  # Construct arch abi flags
+
+  ifeq ($(CONFIG_ARCH_DPFPU),y)
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/Toolchain.defs:108]: why not define ARCHRVISAD correctly at line 83

**Comment by no1wudi** [comment]
Waiting for #5082 , and this patch will be rebased on it.

File: arch/risc-v/Kconfig (hunk 1 above)
Comment (no1wudi) [reply arch/risc-v/Kconfig]: Done, ARCH_RV32IM and ARCH_RV64GC had been removed.

File: boards/risc-v/rv32m1/rv32m1-vega/scripts/Make.defs [hunk 3]
```diff
@@ -49,17 +49,15 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
   ARCHOPTIMIZATION += $(MAXOPTIMIZATION) -fno-strict-aliasing -fno-strength-reduce -fomit-frame-pointer
 endif
 
-ARCHCPURV32IM = -march=rv32imc -mabi=ilp32
-
 ifeq ($(CONFIG_RV32M1_OPENISA_TOOLCHAIN),y)
   ifdef CONFIG_ARCH_RISCV_INTXCPT_EXTREGS
     ifeq ($(filter 0 1 2 3 4 5 , $(CONFIG_ARCH_RISCV_INTXCPT_EXTREGS)),)
       ARCHCPURV32IM = -march=rv32imcxpulpv2
+      ARCHCPUFLAGS = $(ARCHCPURV32IM)
```
Comment (xiaoxiang781216) [inline boards/risc-v/rv32m1/rv32m1-vega/scripts/Make.defs:56]: ARCHCPUFLAGS = -march=rv32imcxpulpv2 and remove ARCHCPURV32IM

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@no1wudi there are some build error need to fix first.

**Comment by no1wudi** [comment]
> @no1wudi there are some build error need to fix first.

Fixed

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by normanr** [comment]
This is giving me:
```
$ make menuconfig
/home/normanr/src/nuttx/nuttx/tools/../Make.defs:23: /home/normanr/src/nuttx/nuttx/tools/../arch/risc-v/src/rv32im/Toolchain.defs: No such file or directory
make: *** No rule to make target '/home/normanr/src/nuttx/nuttx/tools/../arch/risc-v/src/rv32im/Toolchain.defs'.  Stop.
/home/normanr/src/nuttx/nuttx/Make.defs:23: /home/normanr/src/nuttx/nuttx/arch/risc-v/src/rv32im/Toolchain.defs: No such file or directory
make: *** No rule to make target '/home/normanr/src/nuttx/nuttx/arch/risc-v/src/rv32im/Toolchain.defs'.  Stop.
```
Is there anything special I need to do to clear out old state? I just tried `make menuconfig` and `make`.

**Comment by xiaoxiang781216** [comment]
This patch merge Toolchain.defs from rv32im/rv64gc to common, so you may need make distclean and reconfigure again.

**Comment by normanr** [comment]
I couldn't run `make distclean` because of the same reason, also deleting it broke things worse. I had to re-copy the Make.defs from the arch directory, and that fixed things.

**Comment by xiaoxiang781216** [comment]
Yes, copy Mak.defs can fix this problem too.

**Pull request merged**